pkg/controller/controller_utils.go

```diff
@@ -629,6 +686,135 @@ func maxContainerRestarts(pod *api.Pod) int {
+	for i := range pods {
+		pod := pods[i]
+		if !IsPodActive(pod) {
+			glog.V(4).Infof("Ignoring inactive pod %v/%v in state %v, deletion time %v",
+				pod.Namespace, pod.Name, pod.Status.Phase, pod.DeletionTimestamp)
+			continue
+		}
+		controllerRef := getControllerOf(pod.ObjectMeta)
+		if controllerRef != nil {
+			if controllerRef.UID == m.controllerObject.UID {
+				// already controlled
+				if m.controllerSelector.Matches(labels.Set(pod.Labels)) {
+					matchesAndControlled = append(matchesAndControlled, &pod)
+				} else {
+					controlledDoesNotMatch = append(controlledDoesNotMatch, &pod)
+				}
+			} else {
+				// ignoring the pod controlled by other controller
+				glog.V(4).Infof("Ignoring pod %v/%v, it's owned by [%s/%s, name: %s, uid: %s]",
+					pod.Namespace, pod.Name, controllerRef.APIVersion, controllerRef.Kind, controllerRef.Name, controllerRef.UID)
+				continue
+			}
+		} else {
+			if !m.controllerSelector.Matches(labels.Set(pod.Labels)) {
+				continue
+			}
+			matchesNeedsController = append(matchesNeedsController, &pod)
+		}
+	}
+	return matchesAndControlled, matchesNeedsController, controlledDoesNotMatch
+
+}
+
+// getControllerOf returns the controllerRef if controllee has a controller,
+// otherwise returns nil.
+func getControllerOf(controllee api.ObjectMeta) *api.OwnerReference {
+	for _, owner := range controllee.OwnerReferences {
+		// controlled by other controller
+		if owner.Controller != nil && *owner.Controller == true {
+			return &owner
+		}
+	}
+	return nil
+}
+
+// ControlPods sends a patch to take control of the pods. It returns the list of
+// pods that are successfully controlled, and the list of errors that occur in
+// failed patching attempts.
+func (m *podControllerRefManager) ControlPods(pods []*api.Pod) (controlledPods []*api.Pod, errlist []error) {
+	// we should not adopt any pods if the controller is about to be deleted
+	if m.controllerObject.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	for _, pod := range pods {
+		addControllerPatch := fmt.Sprintf(
+			`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true}],"uid":"%s"}}`,
+			m.controllerKind.GroupVersion(), m.controllerKind.Kind,
+			m.controllerObject.Name, m.controllerObject.UID, pod.UID)
+		err := m.podControl.PatchPod(pod.Namespace, pod.Name, []byte(addControllerPatch))
+		// If the pod no longer exists, ignore it.
+		if err != nil && !errors.IsNotFound(err) {
+			errlist = append(errlist, fmt.Errorf("Failed to take control of pod %v/%vby patching its ownerReferences: %v", pod.Namespace, pod.Name, err))
+		} else {
+			controlledPods = append(controlledPods, pod)
+		}
+	}
+	return controlledPods, errlist
+}
+
+// FreePods sends a patch to free the pods from the control of the controller.
+// It returns the list of errors that occur in failed patching attempts.
+func (m *podControllerRefManager) FreePods(pods []*api.Pod) (errlist []error) {
+	for _, pod := range pods {
```

Comment:
You can always make a utility function that does both if you really think it's helpful...

```go
func ApplyToPods(p []*Pod, f func(p *Pod) error) (succeeded, failed []*Pod, errs []errors) { ... }
```